### PR TITLE
[Draft] Testing eBPF for windows programs

### PR DESCRIPTION
**Pull request by vakalapa**
**Reason for Change**:

Experimentation with ebpf-for-windows

**Issue Fixed**:

**Requirements**:

- [ ] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [ ] includes documentation
- [ ] adds unit tests

**Notes**:

### CODE REVIEW

File: test/ebpf/src/endpoint_prog.c [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+        verdict = bpf_map_lookup_elem(policy_map_fd, &key);
+    }
+
+    return (verdict != NULL) ? *verdict : CGROUP_ACT_REJECT;
+}
+
+__inline int
+authorize_v4(bpf_sock_addr_t *ctx, uint8_t direction)
+{
+
```
Comment (dthaler) [inline test/ebpf/src/endpoint_prog.c:79]:
```suggestion
```

File: test/ebpf/src/endpoint_prog.c [hunk 2]
```diff
@@ -0,0 +1,155 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+#include "bpf_helpers.h"
+#include "endpoint_prog.h"
+#include "ebpf_structs.h"
```
Comment (dthaler) [inline test/ebpf/src/endpoint_prog.c:6]:
```suggestion
```
bpf_helpers.h on windows internally includes ebpf_structs.h so you don't need to do it again (and removing it makes your code more cross-plat windows/linux)

File: test/ebpf/testSockProg.c [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+#include "bpf/bpf.h"
+#include "bpf/libbpf.h"
+#include "testSockProg.h"
+
+struct npm_endpoint_prog_t test_ebpf_prog()
+{
+    char *connect4_program_name = "authorize_connect4";
+    char *connect6_program_name = "authorize_connect6";
+    char *recv4_accept_program_name = "authorize_recv_accept4";
+    char *recv6_accept_program_name = "authorize_recv_accept6";
```
Comment (dthaler) [inline test/ebpf/testSockProg.c:12]:
```suggestion
    const char *connect4_program_name = "authorize_connect4";
    const char *connect6_program_name = "authorize_connect6";
    const char *recv4_accept_program_name = "authorize_recv_accept4";
    const char *recv6_accept_program_name = "authorize_recv_accept6";
```

**Review by dthaler — DISMISSED** [review]

File: test/ebpf/testSockProg.c [hunk 4]
```diff
@@ -0,0 +1,115 @@
+
```
[the remark this answers is not in the document]
Comment (vakalapa) [reply test/ebpf/testSockProg.c:1]: i am building in one test VM and moving those binaries across different AKS VMs using this PR, i will clean obj files at the end when i streamline testing process.

File: test/ebpf/src/endpoint_prog.c [hunk 5]
```diff
@@ -65,7 +80,7 @@ _policy_eval(bpf_sock_addr_t *ctx, uint32_t compartment_id, policy_map_key_t key
     if (verdict != NULL)
     {
         char msg[128];
-        bpf_snprintf(msg, sizeof(msg), "Policy Eval: L4 policy ID %d Allowed.", *verdict, sizeof(*verdict));
+        bpf_snprintf(msg, sizeof(msg), "Policy Eval: L4 policy ID %lu Allowed.", (unsigned long *) verdict, sizeof(verdict));
         print_trace(ctx, msg);
```
Comment (dthaler) [inline test/ebpf/src/endpoint_prog.c:84]: Can't you simply do:
```suggestion
        bpf_printk("Policy Eval: L4 policy ID %lu Allowed.", *(unsigned long *) verdict);
```

File: test/ebpf/testSockProg.c [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+    case IP_CACHE_MAP:
+        return IP_CACHE_MAP_NAME;
+    }
+}
+
+int pin_given_map(int internal_map_type, fd_t fd){
+    char *map_name = get_epmap_name(internal_map_type, POLICY_MAP_ID);
+
+    // Map fd is invalid. Open fd to the map.
+    char *pin_path = get_map_pin_path(map_name);
```
Comment (shankarseal) [inline test/ebpf/testSockProg.c:40]: you are leaking `pin_path`

File: test/ebpf/testSockProg.c [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+    }
+
+    printf("get_map_fd: failed to create the MAP\n");
+
+    return INVALID_MAP_FD;
+}
+
+int update_global_policy_map(int compartment_id)
+{
+    fd_t compartment_policy_map_fd = get_map_fd(COMP_POLICY_MAP, 0);
```
Comment (shankarseal) [inline test/ebpf/testSockProg.c:291]: pass `compartment_id` intead of `0` ?

File: test/ebpf/testSockProg.c [hunk 8]
```diff
@@ -0,0 +1,403 @@
...
+}
+
+char *get_epmap_name(int internal_map_type, int comp_id)
+{
+    switch (internal_map_type)
+    {
+    case COMP_POLICY_MAP:
+    {
+        int map_name_size = (sizeof(char) * strlen(COMP_PMAP_NAME_PREFIX)) + sizeof(int);
+        char full_map_name[map_name_size];
```
Comment (shankarseal) [inline test/ebpf/testSockProg.c:25]: You have to allocate on heap.

File: test/ebpf/testSockProg.c [hunk 9]
```diff
@@ -0,0 +1,428 @@
...
+    printf("recv program\n");
+
+    struct bpf_program *recv6_accept_program = bpf_object__find_program_by_name(object, recv6_accept_program_name);
+    if (!recv6_accept_program)
+    {
+        printf("%s is null\n", recv6_accept_program_name);
+        return _npm_endpoint_prog_t;
+    }
+
+    _npm_endpoint_prog_t.recv4_accept_program = recv6_accept_program;
```
Comment (shankarseal) [inline test/ebpf/testSockProg.c:133]: bug

File: test/ebpf/testSockProg.c [hunk 10]
```diff
@@ -0,0 +1,428 @@
...
+#include <stdlib.h>
+#include <string.h>
+#include "bpf/bpf.h"
+#include "bpf/libbpf.h"
+#include "testSockProg.h"
+
+char *get_map_pin_path(const char *map_name)
+{
+    char *map_pin_path = malloc(sizeof(char) * (strlen(map_name) + strlen(DEFAULT_MAP_PIN_PATH_PREFIX) + 1));
+    strcpy(map_pin_path, DEFAULT_MAP_PIN_PATH_PREFIX);
```
Comment (dthaler) [inline test/ebpf/testSockProg.c:11]: check for malloc failure?  same in line 25

File: test/ebpf/testSockProg.c [hunk 11]
```diff
@@ -0,0 +1,428 @@
...
+        return 0;
+    }
+
+    printf("pin_given_map: pinned map not found, creating pin for map %s\n", map_name);
+    // Map created. Now pin the map.
+    int error = bpf_obj_pin(fd, pin_path);
+    if (error != 0)
+    {
+        // close map fd.
+        _close(fd);
```
Comment (dthaler) [inline test/ebpf/testSockProg.c:61]: This looks like a bug.  One should not call _close on FD's returned from `bpf_map__fd()`.  Instead one should call `bpf_object__close()` on the object returned by `bpf_prog_load()` when done.   `bpf_map__fd()` just returns an FD already stored in the object.  The docs on https://microsoft.github.io/ebpf-for-windows/libbpf_8h.html#a37b9672527f744e7492797529f4f9a6a make this clear for the fd returned by bpf_prog_load but we can update the docs to be clear about the fds returned by bpf_map_fd too.

File: test/ebpf/testSockProg.c [hunk 12]
```diff
@@ -0,0 +1,428 @@
...
+    }
+
+    _npm_endpoint_prog_t.object = object;
+
+    printf("Getting the bpf_prog for connect program\n");
+    struct bpf_program *connect_program = bpf_object__find_program_by_name(object, connect4_program_name);
+    if (!connect_program)
+    {
+        printf("%s is null\n", connect4_program_name);
+        return _npm_endpoint_prog_t;
```
Comment (dthaler) [inline test/ebpf/testSockProg.c:100]: here and below through the rest of this function, this leaks object.  Should call [bpf_object__close(object)](https://microsoft.github.io/ebpf-for-windows/libbpf_8h.html#a473872ff7ab53d1ed593e7abd206f31a) when done with it.

File: test/ebpf/testSockProg.c [hunk 13]
```diff
@@ -0,0 +1,428 @@
...
+    printf("Now getting MAP fds and pinning them\n");
+
+    struct bpf_map *map_policy_maps_obj = bpf_object__find_map_by_name(object, "map_policy_maps");
+    if (map_policy_maps_obj == NULL)
+    {
+        printf("global policy map is null\n");
+        return _npm_endpoint_prog_t;
+    }
+
+    int err = pin_given_map(GLOBAL_POLICY_MAP, bpf_map__fd(map_policy_maps_obj));
```
Comment (dthaler) [inline test/ebpf/testSockProg.c:144]: if you just want to pin all maps in the object, you can just call [bpf_object__pin_maps()](https://microsoft.github.io/ebpf-for-windows/libbpf_8h.html#ae1bffc48512a68600518c0f5d516b87e)

File: test/ebpf/testSockProg.c [hunk 14]
```diff
@@ -0,0 +1,428 @@
...
+    if (compartment_policy_map_fd == INVALID_MAP_FD)
+    {
+        return -1;
+    }
+
+    printf("Updating comp policy map\n");
+    fd_t global_policy_map_fd = get_map_fd(GLOBAL_POLICY_MAP, 0);
+    if (global_policy_map_fd == INVALID_MAP_FD)
+    {
+        return INVALID_MAP_FD;
```
Comment (dthaler) [inline test/ebpf/testSockProg.c:345]: here and below leaks compartment_policy_map_fd

File: test/ebpf/testSockProg.c [hunk 15]
```diff
@@ -0,0 +1,428 @@
...
+    }
+
+    // Update the global policy map
+    int error = bpf_map_update_elem(global_policy_map_fd, &compartment_id, &compartment_policy_map_fd, BPF_ANY);
+    if (error != 0)
```
Comment (dthaler) [inline test/ebpf/testSockProg.c:350]: leaks global_policy_map_fd

File: test/ebpf/testSockProg.c [hunk 16]
```diff
@@ -0,0 +1,428 @@
...
+        .remote_port = remote_port,
+    };
+
+    if (!delete)
+    {
+        int result = bpf_map_update_elem(comp_policy_map_fd, &key, &policy_id, 0);
+        if (result != 0)
+        {
+            printf("Error while updating comp policy map\n");
+            return result;
```
Comment (dthaler) [inline test/ebpf/testSockProg.c:379]: here and below leaks comp_policy_map_fd

File: test/ebpf/testSockProg.c [hunk 17]
```diff
@@ -0,0 +1,428 @@
...
+    ip_address_t ip_cache_key = {0};
+    ip_cache_key.ipv4 = ipv4;
+
+    if (!delete)
+    {
+        int result = bpf_map_update_elem(ip_cache_map_fd, &ip_cache_key, &ctx_label_id, 0);
+        if (result != 0)
+        {
+            printf("Error while updating ip cache map %d\n", result);
+            return result;
```
Comment (dthaler) [inline test/ebpf/testSockProg.c:413]: here and below leaks ip_cache_map_fd

File: test/ebpf/testSockProg.c (hunk 11 above)
Comment (dthaler) [reply test/ebpf/testSockProg.c:61]: FYI, documentation has now been updated

File: test/ebpf/main.go [hunk 18]
```diff
@@ -48,13 +48,10 @@ func main() {
 	}
 
 	// 85-testing cluster
-	// block all traffic on pod with IP 10.240.0.37
-	// which is win-webserver-7b7d755975-hxswl with
-	// "ID":  "2e931103-8088-4873-b0fa-f5f8adc13c3b",
-	// "Name":  "35a1fa7e-eth0",
-	// "CompartmendId":  4,
+	// block all traffic on pod with IP 10.240.0.47
+	// "CompartmendId":  3,
```
Comment (dthaler) [inline test/ebpf/main.go:52]:
```suggestion
	// "CompartmentId":  3,
```

File: test/ebpf/src/endpoint_prog.c [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+    .value_size = sizeof(uint32_t),       // value is policy ID
+    .max_entries = POLICY_MAP_SIZE,
+    .id = POLICY_MAP_ID};
+
+// define the outer map of policy map policy map
+SEC("maps")
+struct bpf_map_def map_policy_maps = {
+    .type = BPF_MAP_TYPE_ARRAY_OF_MAPS,
+    .key_size = sizeof(uint32_t),   // key is a compartment ID
+    .value_size = sizeof(uint32_t), // value is FD of the policy map specific to that compartment
```
[the remark this answers is not in the document]
Comment (vakalapa) [reply test/ebpf/src/endpoint_prog.c:36]: Okay, will correct the comment.

File: test/ebpf/main.go [hunk 20]
```diff
@@ -0,0 +1,158 @@
...
+
+	return state, 0
+}
+
+func test_scenario(compID int) int {
+
+	iptoid := map[string]uint32{
+		"10.240.0.16": 123,
+		"10.240.0.41": 456,
+		"10.240.0.15": 789,
```
[the remark this answers is not in the document]
Comment (vakalapa) [reply test/ebpf/main.go:96]: Today Azure CNI supports only ipv4, and this test AKS-Engine cluster has only IPv4 addresses attached. So i am testing only v4 for time being. Once the basic functionality is done probably i will try adding v6 addrs.

File: test/ebpf/testSockProg.c (hunk 13 above)
Comment (vakalapa) [reply test/ebpf/testSockProg.c:144]: Awesome will try this.

File: test/ebpf/testSockProg.c (hunk 11 above)
Comment (vakalapa) [reply test/ebpf/testSockProg.c:61]: Okay, will remove this call.

File: test/ebpf/testSockProg.c [hunk 21]
```diff
@@ -0,0 +1,437 @@
...
+    }
+
+    printf("pin_given_map: pinned map not found, creating pin for map %s\n", map_name);
+    // Map created. Now pin the map.
+    int error = bpf_obj_pin(fd, pin_path);
+    if (error != 0)
+    {
+        return -1;
+    }
+    return 0;
```
Comment (dthaler) [inline test/ebpf/testSockProg.c:79]: This leaks fdnew

File: npm/pkg/controlplane/controllers/ebpf/namespaceController.go [hunk 22]
```diff
@@ -0,0 +1,476 @@
...
+	"k8s.io/klog"
+)
+
+type LabelAppendOperation bool
+
+const (
+	clearExistingLabels    LabelAppendOperation = true
+	appendToExistingLabels LabelAppendOperation = false
+)
+
```
Comment (matmerr) [inline npm/pkg/controlplane/controllers/ebpf/namespaceController.go:34]: delete this file

**Comment by dthaler** [comment]
@vakalapa should this PR be reopened?

**Pull request closed without merging**